metal/cloud.go

```diff
@@ -28,36 +24,17 @@ const (
 	checkLoopTimerSeconds        = 60
 )
 
-type nodeReconciler func(ctx context.Context, nodes []*v1.Node, mode UpdateMode) error
-type serviceReconciler func(ctx context.Context, services []*v1.Service, mode UpdateMode) error
-
-// cloudService an internal service that can be initialize and report a name
-type cloudService interface {
+type inittable interface {
 	name() string
-	init(k8sclient kubernetes.Interface) error
-	nodeReconciler() nodeReconciler
-	serviceReconciler() serviceReconciler
-}
-
-type cloudInstances interface {
-	cloudprovider.InstancesV2
-	cloudService
-}
-type cloudLoadBalancers interface {
-	cloudprovider.LoadBalancer
-	cloudService
-}
-type cloudZones interface {
-	cloudprovider.Zones
-	cloudService
+	init(k8sclient kubernetes.Interface, stop <-chan struct{}) error
 }
 
 // cloud implements cloudprovider.Interface
 type cloud struct {
 	client                      *packngo.Client
-	instances                   cloudInstances
-	zones                       cloudZones
-	loadBalancer                cloudLoadBalancers
+	instances                   *instances
```

Comment:
Are we relying on type implementation details in this PR? What's the motivation for moving away from interface typed fields here? Could we have this, for example:
```
instances cloudprovider.InstancesV2
zones cloudprovider.Zones
loadBalancer cloudprovider.LoadBalancer
```

I believe @detiber suggested that zones could be entirely removed based on his work in https://github.com/equinix/cloud-provider-equinix-metal/compare/master...detiber:test?expand=1

Reply:
Yeah, that is a good point.

I cannot use the `cloud provider.InstancesV2` etc. interfaces because these types also implement `interface inittable`. That gets used here:

```go
func (c *cloud) Initialize(clientBuilder cloudprovider.ControllerClientBuilder, stop <-chan struct{}) {
	klog.V(5).Info("called Initialize")
	clientset := clientBuilder.ClientOrDie("cloud-provider-equinix-metal-shared-informers")
	// pass the clientset and stop channel to each service that needs it
	services := c.services()
	for _, elm := range services {
		if err := elm.init(clientset, stop); err != nil {
			klog.Fatalf("could not initialize %s: %v", elm.name(), err)
		}
	}
```

When we call `newCloud`, we do not yet have the k8s client or the stop channel, which some of these need to initialize.

If we could restructure these, it would make a lot of that go away, and we would not need anything other than the standard interfaces. I will see if I can do that.

Thanks for putting the effort in to look at it in depth @displague .

Reply:
Success. I managed to put the creation of the objects inside `cloud.Initialize()`, which allowed me to pass everything, eliminating the need for a separate `init()`, and therefore `interface inittable`.

This simplified many things, so I am happy. 

I did a cleanup of lots of lint checks while I was at it.

Reply:
Got it.